keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -323,11 +339,171 @@ myOtherField: newly added field
+validation is applied to the object owning the property. Thus, if a new field
+becomes required the object cannot have other fields added onto it or changed
+without also adding the newly required field.
+
+Kubernetes has precedent for using a weaker form of racheting when validation is
+strengthened:
+
+```go
+if !Validation(old) {
+  // Ignore errors caused by !Validation(new)
+}
+```
+
+An example of this is the structural schema requirement added when Kubernetes
+went GA ([source](https://github.com/kubernetes/kubernetes/blob/12dc19d46fb97cbbfeb1e12b8a10ff7ae73d9515/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L1583-L1590)): 
+```go
+// requireStructuralSchema returns true if schemas specified must be structural
+func requireStructuralSchema(oldCRDSpec *apiextensions.CustomResourceDefinitionSpec) bool {
+	if oldCRDSpec != nil && specHasNonStructuralSchema(oldCRDSpec) {
+		// don't tighten validation on existing persisted data
+		return false
+	}
+	return true
+}
+```
+
+It is not possible to generically allow this weaker form of ratcheting on all
+types of validations - it may allow data that would have failed a validation
+on an older schema and break controller workflows.
+
+To enable schemas to add validations with more complicated ratcheting logic
+than a DeepEqual check (e.g. to allow a new `minLength` restriction to allow any
+value as long as it is getting closer the new minimum); we turn to CEL.
+
+#### Ratcheting Rules in CEL
+
+CRD Validation Rules currently have two types:
+
+- Normal Rule (Applies to  all values)
+- Transition Rule (Applies only on UPDATE in places where old could be correlated to new)
+
+With only these rule types we are not able to express any logic of the form 
+"old values must obey one schema, and new values another". We propose to add a 
+new change to `ValidationRule` to enable this functionality for ratcheting:
+
+```go
+// ValidationRule describes a validation rule written in the CEL expression language.
+type ValidationRule struct {
+  // ...
+	Rule string `json:"rule" protobuf:"bytes,1,opt,name=rule"`
+	Message string `json:"message,omitempty" protobuf:"bytes,2,opt,name=message"`
+	MessageExpression string `json:"messageExpression,omitempty" protobuf:"bytes,3,opt,name=messageExpression"`
+	Reason *FieldValueErrorReason `json:"reason,omitempty" protobuf:"bytes,4,opt,name=reason"`
+	FieldPath string `json:"fieldPath,omitempty" protobuf:"bytes,5,opt,name=fieldPath"`
+  // ...
+
+  // New
+  // Normally, if `rule` references `oldSelf` then it is skipped on create, and
+  // when an old value could not be found. If this field is true then the rule is 
+  // always evaluated. `oldSelf = nil` on create or when an old value could not 
+  // be correlated.
+  //
+  // Ensure your rule employs a nil check on `oldSelf` before using it in an
+  // expression.
+  AppliesOnCreate bool `json:"requiresOldValue"`
```

Comment:
AppliesOnCreate or AppliesOnCreate ?

Reply:
```suggestion
  AppliesOnCreate bool `json:"appliesOnCreate"`
```

Reply:
`requiresOldValue` to me means that `oldSelf` can't be nil, which is not backwards compatible (since default behavior would be `true`). Will defer naming to API review

Reply:
was more about the mismatch between field name and json name.